hazelcast/src/main/java/com/hazelcast/spi/impl/SlowOperationDetector.java

```diff
@@ -0,0 +1,214 @@
+/*
+ * Copyright (c) 2008-2013, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.spi.impl;
+
+import com.hazelcast.instance.GroupProperties;
+import com.hazelcast.instance.HazelcastThreadGroup;
+import com.hazelcast.logging.ILogger;
+import com.hazelcast.logging.Logger;
+import com.hazelcast.spi.impl.operationexecutor.OperationRunner;
+import com.hazelcast.util.ConcurrencyUtil;
+import com.hazelcast.util.ConstructorFunction;
+import com.hazelcast.util.EmptyStatement;
+
+import java.util.Collection;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.TimeUnit;
+
+import static java.lang.String.format;
+
+/**
+ * Monitors the operation threads to find and collect data of long running operations.
+ */
+final class SlowOperationDetector {
+
+    private static final int FULL_LOG_FREQUENCY = 100;
+    private static final long ONE_SECOND_IN_NANOS = TimeUnit.SECONDS.toNanos(1);
+    private static final ILogger LOGGER = Logger.getLogger(SlowOperationDetector.class);
+
+    private final ConcurrentHashMap<Integer, SlowOperationLog> slowOperationLogs
+            = new ConcurrentHashMap<Integer, SlowOperationLog>();
+    private final StringBuilder stackTraceStringBuilder = new StringBuilder();
+
+    private final OperationRunner[] genericOperationRunners;
+    private final OperationRunner[] partitionOperationRunners;
+
+    private final CurrentOperationData[] genericCurrentOperationData;
+    private final CurrentOperationData[] partitionCurrentOperationData;
+
+    private final long slowOperationThresholdNanos;
+    private final long logPurgeIntervalNanos;
+    private final long logRetentionNanos;
+
+    private volatile boolean running = true;
+
+    public SlowOperationDetector(OperationRunner[] genericOperationRunners,
+                                 OperationRunner[] partitionOperationRunners,
+                                 GroupProperties groupProperties, HazelcastThreadGroup hazelcastThreadGroup) {
+        this.genericOperationRunners = genericOperationRunners;
+        this.partitionOperationRunners = partitionOperationRunners;
+
+        genericCurrentOperationData = initCurrentOperationData(genericOperationRunners);
+        partitionCurrentOperationData = initCurrentOperationData(partitionOperationRunners);
+
+        slowOperationThresholdNanos = getGroupPropertyMillis(groupProperties.SLOW_OPERATION_DETECTOR_THRESHOLD_MILLIS);
```

Comment:
Bit confused here because you are gettingXXXMillies but actually they are nanos. Perhaps getMilliesToNanos(...)

This way it is clear that you are expecting to find a millies and converting it to nanos. Anyhow.. not extremely relevant.